packages/trackerless-network/src/NetworkStack.ts

```diff
@@ -0,0 +1,149 @@
+
+    private connectionManager?: ConnectionManager
+    private layer0DhtNode?: DhtNode
+    private streamrNode?: StreamrNode
+    private readonly metricsContext: MetricsContext
+    private readonly options: NetworkOptions
+    private stopped = false
+    private readonly firstConnectionTimeout: number
+    private dhtJoinRequired = true
+
+    constructor(options: NetworkOptions) {
+        super()
+        this.options = options
+        this.metricsContext = options.metricsContext || new MetricsContext()
+        this.layer0DhtNode = new DhtNode({
+            ...options.layer0,
+            metricsContext: this.metricsContext
+        })
+        this.streamrNode = new StreamrNode({
+            ...options.networkNode,
+            nodeName: options.networkNode.nodeName || options.layer0.nodeName,
+            metricsContext: this.metricsContext
+        })
+        this.firstConnectionTimeout = options.networkNode.firstConnectionTimeout || 5000
+    }
+
+    async start(doJoin = true): Promise<void> {
+        await this.layer0DhtNode!.start()
+        this.connectionManager = this.layer0DhtNode!.getTransport() as ConnectionManager
+        const entryPoint = this.options.layer0.entryPoints![0]
+        if (isSamePeerDescriptor(entryPoint, this.layer0DhtNode!.getPeerDescriptor())) {
+            this.dhtJoinRequired = false
+            await this.layer0DhtNode?.joinDht(entryPoint)
+            await this.streamrNode?.start(this.layer0DhtNode!, this.connectionManager!, this.connectionManager!)
+        } else {
+            if (doJoin) {
+                this.dhtJoinRequired = false
+                await this.joinDht()
+            }
+            await this.streamrNode?.start(this.layer0DhtNode!, this.connectionManager!, this.connectionManager!)
+        }
+    }
+
+    private async joinDht(): Promise<void> {
+        setImmediate(() => {
+            this.layer0DhtNode?.joinDht(this.options.layer0.entryPoints![0])
+        })
+        await this.waitForFirstConnection()
+    }
+
+    private async waitForFirstConnection(): Promise<void> {
+        const readynessListener = new ReadynessListener(this, this.layer0DhtNode!)
+        await readynessListener.waitUntilReady(this.firstConnectionTimeout)
+    }
+
+    async joinLayer0IfRequired(streamPartId: StreamPartID): Promise<void> {
+        if (this.isJoinRequired(streamPartId)) {
+            this.dhtJoinRequired = false
+            await this.joinDht()
+        } else if (this.layer0DhtNode!.getNumberOfConnections() < 1) {
+            await this.waitForFirstConnection()
+        }
+    }
+
+    private isJoinRequired(streamPartId: StreamPartID): boolean {
+        return this.dhtJoinRequired && !this.layer0DhtNode!.hasJoined() && this.streamrNode!.isJoinRequired(streamPartId)
+    }
+
+    getStreamrNode(): StreamrNode {
+        return this.streamrNode!
+    }
+
+    getConnectionManager(): ConnectionManager | undefined {
```

Comment:
This method is not used, could remove? Also similar issue e.g. `DhtNode#hasJoined` and `Transport#getAllConnectionPeerDescriptors`. 

IntellijJ's "Inspect code" tool finds these quite easily (the "Unused symbols" section of the report)

Reply:
Removed `getConnectionManager` the other two are used in code.